Create live payload handler

jkonecny12 opened the pull request:
Add Live payload handler.
Add logic to the payload main module to work with multiple payloads.
Migrate old tests for this new solution.
Add test for the new payload handler.

jkonecny12 commented on pyanaconda/modules/payload/live/live.py:
```diff
@@ -0,0 +1,126 @@
+log = get_module_logger(__name__)
+
+
+class LiveHandlerModule(KickstartBaseModule):
```
I would like to decide this in another PR. I'm also considering option to merge these two but I don't want to decide anything before.

It is not a problem to rename it later.

poncovka replied:
Well, it is not just a name of this module. You will have to rename the namespace, the object identifier, the interface class, the directory and the tests. It will affect almost every file in this pull request, so I would prefer to have it decided now.

poncovka replied:
Then I would go the other way. Let's have separate objects for these payloads and merge them eventually later. It looks like an extra work with uncertain results.

jkonecny12 replied:
After our discussion we picked names LiveOS and LiveImage for handlers.

jkonecny12 commented on pyanaconda/modules/common/constants/namespaces.py:
```diff
@@ -71,6 +71,11 @@
     "DNF"
 )
 
+LIVE_NAMESPACE = (
```
Fixed.

jkonecny12 commented on pyanaconda/modules/payload/live/live.py:
```diff
@@ -0,0 +1,126 @@
+        liveimg = data.liveimg
+
+        log.debug("Live object: %s", liveimg)
+        log.debug("Live parts: %s", liveimg.url)
```
Ups. Fixed.

jkonecny12 commented on pyanaconda/modules/payload/live/live.py:
```diff
@@ -0,0 +1,126 @@
+        """
+        return self._url
+
+    def _set_url(self, url):
```
Why do you think so? Do we have a reason to change this other way than a kickstart file?

poncovka replied:
It should be possible to use this module without a kickstart file. It is easy to implement the DBus setters, so why not?

jkonecny12 replied:
I don't like the idea extending our DBus API when it is not required but yeah I agree in this case it will make our testing and debugging.

jkonecny12 commented on pyanaconda/modules/payload/live/live.py:
```diff
@@ -0,0 +1,126 @@
+        """Setup the kickstart data."""
+        liveimg = data.liveimg
+
+        liveimg.url = self._url
```
Fixed.

jkonecny12 commented on pyanaconda/modules/payload/live/live_interface.py:
```diff
@@ -0,0 +1,57 @@
+    def connect_signals(self):
+        super().connect_signals()
+
+        self.implementation.url_changed.connect(self.changed("Url"))
```
Fixed.

jkonecny12 commented on pyanaconda/modules/payload/payload.py:
```diff
@@ -50,8 +53,18 @@ def kickstart_specification(self):
+        if data.liveimg.seen:
+            self._payload_handler = LiveHandlerModule()
+        else:
+            self._payload_handler = DNFHandlerModule()
```
Yeah I know but I don't want to complicate things for now.

poncovka replied:
Ok.

jkonecny12 commented on tests/nosetests/pyanaconda_tests/module_payload_shared.py:
```diff
@@ -0,0 +1,43 @@
+from pyanaconda.modules.payload.payload import PayloadModule
+
+
+class PayloadHandlerBase(unittest.TestCase):
```
Having all the kickstart tests for all the handlers in one file would be overwhelming. I want to avoid that if possible and I also want to avoid copy-paste code.

Yeah it will be evaluated but when there are no tests I don't see a problem. The other possible solution is to have this as a mixin class. I don't have problem with that but the separate file for this is still required. This file can be moved to another directory if you want?

poncovka replied:
Ok, create at least a mixin class, please.

jkonecny12 replied:
Fixed.

jkonecny12 commented on pyanaconda/modules/payload/payload.py:
```diff
@@ -50,8 +53,18 @@ def kickstart_specification(self):
 
+    def _create_correct_handler(self, data):
+        if data.liveimg.seen:
+            self._payload_handler = LiveHandlerModule()
```
I don't want to make this PR bigger. I want to change how the payload module works with handlers but in separate PR.

poncovka replied:
Ok.

jkonecny12 commented on pyanaconda/modules/payload/payload.py:
```diff
@@ -50,8 +53,18 @@ def kickstart_specification(self):
+        else:
+            self._payload_handler = DNFHandlerModule()
+
     def generate_kickstart(self):
```
Same as above. It won't matter for now because it is not used and I don't think it should be part of this PR.

poncovka replied:
I disagree, but ok.

poncovka replied:
Btw. you should have a test for this case.

jkonecny12 commented on pyanaconda/modules/payload/payload.py:
```diff
@@ -50,8 +53,18 @@ def kickstart_specification(self):
+        self._create_correct_handler(data)
+        self._publish_handler()
+
         self._payload_handler.process_kickstart(data)
```
This whole solution will probably change so I don't see a reason to fix this now.

poncovka replied:
Ok.

jkonecny12 commented:
**UPDATED:**
* Renamed `Live` to `LiveImage` there will be another one `LiveOS` in the same directory (other PR).
* Added setters.
* Fixed tests.
* Added setters tests.
* Base class for payload tests is now mixin.

jkonecny12 commented:
**UPDATED:**
* Minor comment fix.

jkonecny12 commented:
jenkins, test this please
